kolibri/content/management/commands/importcontent.py

```diff
@@ -147,34 +149,42 @@ def _transfer(self, method, channel_id, path=None, node_ids=None, exclude_node_i
                     srcpath = paths.get_content_storage_file_path(filename, datafolder=path)
                     filetransfer = transfer.FileCopy(srcpath, dest)
 
-                try:
-
-                    with filetransfer:
-
-                        with self.start_progress(total=filetransfer.total_size) as file_dl_progress_update:
-
-                            for chunk in filetransfer:
-                                if self.is_cancelled():
-                                    filetransfer.cancel()
-                                    break
-                                length = len(chunk)
-                                overall_progress_update(length)
-                                file_dl_progress_update(length)
-
-                    file_checksums_to_annotate.append(f.id)
-
-                except HTTPError:
-                    overall_progress_update(f.file_size)
-
-                except OSError:
-                    number_of_skipped_files += 1
-                    overall_progress_update(f.file_size)
+                finished = False
+                while not finished:
+                    try:
+                        with filetransfer:
+                            with self.start_progress(total=filetransfer.total_size) as file_dl_progress_update:
+                                for chunk in filetransfer:
+                                    if self.is_cancelled():
+                                        filetransfer.cancel()
+                                        finished = True
+                                        break
+                                    length = len(chunk)
+                                    overall_progress_update(length)
+                                    file_dl_progress_update(length)
+
+                        file_checksums_to_annotate.append(f.id)
+                        finished = True
+
+                    # When there is an Internet connection error, wait for 30 seconds to retry
+                    except ConnectionError:
+                        for i in range(30):
+                            if self.is_cancelled():
+                                self.cancel()
+                                finished = True
+                                break
+                            sleep(1)
+
+                    # When the file cannot be found on the server or disk, skip it
+                    except (HTTPError, OSError):
```

Comment:
This is skipping files on _all_ HTTP errors, not just `404`s. Some (e.g. `503`) should trigger a retry, similar to a timeout. Other `4xx` and `5xx` errors should probably cause an actual import failure.

Reply:
cc @aronasorman you'd probably have a better sense for how we want to handle various error codes that might be [returned from cloudflare](https://support.cloudflare.com/hc/en-us/articles/115003011431/#cloudflare_specific_codes)

maybe `502`, `503`, and `504` should all trigger the retry logic?

Reply:
I updated the code to catch the `Timeout` error as well so that it should cover the errors with status codes such as `522`, `524`. I think other errors with the status codes that trigger the retry logic should be covered by `ConnectionError`. So when the `HTTPError` is caught, it should only get the errors that don't trigger the retry logic?

Reply:
There are three possible outcomes when we encounter either `HTTPError` or `OSError`:

1. _retry_ the same file again
2. _skip_ the file and try the next one
3. _fail_ the entire transfer

| Exception | Condition | Outcome |
|--|--|--|
| `HTTPError` | `e.code == 404` | skip |
| `HTTPError` | `e.code in (502, 503, 504, 521, 522, 523, 524)` | retry |
| `HTTPError` | `else` | fail |
| `OSError` | `e.errno == errno.ENOENT` | skip |
| `OSError` | `else` | fail |

Reply:
> I updated the code to catch the Timeout error as well so that it should cover the errors with status codes such as 522, 524. I think other errors with the status codes that trigger the retry logic should be covered by ConnectionError. So when the HTTPError is caught, it should only get the errors that don't trigger the retry logic?

I'm having trouble finding documentation which says that 5xx error codes raise `Timeout` or `ConnectionError`. If so, that's great! Do you have references?

Reply:
Thank you for the table! It's very clear information.
I didn't find documentation that explains specific error codes for different exceptions, but I found [this doc](https://github.com/requests/requests/blob/v2.18.4/docs/user/quickstart.rst#errors-and-exceptions) that somewhat explains what errors cause `ConnectionError` and `Timeout`. It is my assumption that the errors mentioned in the doc above are related to certain error codes...

Reply:
would it be possible to write tests which emulate the five scenarios above and ensure that the expected behavior happens?

Reply:
hmm I'm not sure about emulate the second scenario, but I can try with the other four first!

Reply:
yeah I believe you should be able to mock the `filetransfer` object with some dummy code that raises variations of the exceptions listed in the table? I think @jamalex wrote that wrapper so he might have some pointers

Reply:
I think I find the way to test the second scenario now, will update the tests soon!

Reply:
Since the filetransfer object uses `requests.get`, the most robust method might be to mock that to return those status codes, and then see how the error bubbles up to your code.